cli/lsp/testing.rs

```diff
@@ -0,0 +1,1959 @@
+        let mut maybe_name = None;
+        let mut steps = None;
+        for prop in &obj_lit.props {
+          if let ast::PropOrSpread::Prop(prop) = prop {
+            match prop.as_ref() {
+              ast::Prop::KeyValue(key_value_prop) => {
+                if let ast::PropName::Ident(ast::Ident { sym, .. }) =
+                  &key_value_prop.key
+                {
+                  match sym.to_string().as_str() {
+                    "name" => match key_value_prop.value.as_ref() {
+                      // matches string literals (e.g. "test name" or
+                      // 'test name')
+                      ast::Expr::Lit(ast::Lit::Str(lit_str)) => {
+                        maybe_name = Some(lit_str.value.to_string());
+                      }
+                      // matches template literals with only a single quasis
+                      // (e.g. `test name`)
+                      ast::Expr::Tpl(tpl) => {
+                        if tpl.quasis.len() == 1 {
+                          if let Some(tpl_element) = tpl.quasis.get(0) {
+                            maybe_name =
+                              Some(tpl_element.raw.value.to_string());
+                          }
+                        }
+                      }
+                      _ => (),
+                    },
+                    "fn" => match key_value_prop.value.as_ref() {
+                      ast::Expr::Arrow(arrow_expr) => {
+                        steps = arrow_to_steps(parent, level, arrow_expr);
+                      }
+                      ast::Expr::Fn(fn_expr) => {
+                        steps = fn_to_steps(parent, level, &fn_expr.function);
+                      }
+                      _ => (),
+                    },
+                    _ => (),
+                  }
+                }
+              }
+              ast::Prop::Method(method_prop) => {
+                steps = fn_to_steps(parent, level, &method_prop.function);
+              }
+              _ => (),
+            }
+          }
+        }
+        maybe_name.map(|name| (name, steps))
+      }
+      ast::Expr::Fn(fn_expr) => {
+        if let Some(ast::Ident { sym, .. }) = fn_expr.ident.as_ref() {
+          let name = sym.to_string();
+          let steps = fn_to_steps(parent, level, &fn_expr.function);
+          Some((name, steps))
+        } else {
+          None
+        }
+      }
+      ast::Expr::Lit(ast::Lit::Str(lit_str)) => {
+        let name = lit_str.value.to_string();
+        let mut steps = None;
+        match node.args.get(1).map(|es| es.expr.as_ref()) {
+          Some(ast::Expr::Fn(fn_expr)) => {
+            steps = fn_to_steps(parent, level, &fn_expr.function);
+          }
+          Some(ast::Expr::Arrow(arrow_expr)) => {
+            steps = arrow_to_steps(parent, level, arrow_expr);
+          }
+          _ => (),
+        }
+        Some((name, steps))
+      }
+      _ => None,
+    }
+  } else {
+    None
+  }
+}
```

Comment:
This file is very large. We might want to consider splitting it up into separate modules in an `lsp/testing` folder (ex. this code analysis code could maybe be in its own file).

Reply:
Done.